### ISSUE TITLE
It's not possible to store IDN domain names in customer data

### ISSUE THREAD

**Issue by AndreasSt-2022** [Dolibarr/dolibarr#23557]
### Bug

When entering an internationalized domain name in the web address or email address of a customer the special characters are removed and a totally wrong domain name is stored.

### Environment Version

probably all, at least 16.0.3 and 11.0.3

### Environment OS

Gentoo Linux, probably others as well

### Environment Web server

nginx, probably others as well

### Environment PHP

7.4.33 and others

### Environment Database

MariaDB-10.5.16, probably others as well

### Expected and actual behavior

When entering the data for a new customer and using a domain name like français.example.com or anhänger.example.com in the web address or email address field the special characters get removed. The stored domain name is then franais.example.com or anhnger.example.com which is of course totally wrong.

Expected behavior would be that the domain name stays as entered.

If that is not possible there should at least be an error message. Storing a wrong domain name without any comment is a bad idea.

### Steps to reproduce the behavior

Enter a non-ASCII character in the web address or email address field of a business partner and save the changed data.

**Comment by AndreasSt-2022** [comment]
The dolibarr demo at demo.dolibarr.biz (version 16.0.0) shows the same behavior.

**Comment by hasse-basse** [comment]
I belived it could be a charset setting in the DB so I made a test.

I created a new client and wrote the domain name åäö.com. It resulted in that åäö was delted and just .com remaind. I had a look in the DB and there as well was just .com.

I then wrote åäö.com in the DB and then I could read åäö.com both in the DB and on the client card in Dolibarr!

From that I think we can draw the conclusion that there is a filter in the Dolibarrr code preventing us to use special characters! I agree with you that it should be taken away as today it is possible to use special characters in the URL. Anyhow, I don't know how to do this so I have to leave it to someone else.

**Comment by hasse-basse** [comment]
I've figured out where the code is but I don't know how to change it

societe/card.php line 468

**Comment by AndreasSt-2022** [comment]
The bug is still present in version 18.0.3.

If you don't want to store the special characters in the database the name should at least be converted to punycode. Then it will look ugly but it will be working.

**Comment by AndreasSt-2022** [comment]
The bug is still present in version 20.0.3.

**Comment by JRFPlows** [comment]
Still present in 20.0.4, i only had one customer and changed "llx_societe email url" manually with putty and mysql, worked perfect

**Comment by hasse-basse** [comment]
Yeah, as I wrote in January 2023, it is a filter in the php code, it is 5 minutes work for someone who know how to make a pull request.

**Comment by hregis** [comment]
@hasse-basse @JRFPlows what you understand with : changed "llx_societe email url" manually ?

**Comment by hasse-basse** [comment]
If I write special characters directly into the database, it works without any issues. However, the Dolibarr interface refuses special characters in input but if they have been edited in the DB they will appear in the Dolibarr interface.

It has been two years since I last looked into this, but as I remember, I found a character filter in the PHP code that prevents typing, for example, å, ä, and ö. Back in 2023, its location was here:  societe/card.php line 468

**Comment by JRFPlows** [comment]
> [@hasse-basse](https://github.com/hasse-basse) [@JRFPlows](https://github.com/JRFPlows) what you understand with : changed "llx_societe email url" manually ?

I changed the content in the table llx_societe two fields "email" and "url" for one customer manually in mysql

so the filter "filter_sanitize_email" and "filter_sanitize_url" could not change my input.

i have not found a way to tell the functions to allow special characters such as öüä, as the functions are within PHP.

**Comment by JRFPlows** [comment]
> If I write special characters directly into the database, it works without any issues. However, the Dolibarr interface refuses special characters in input but if they have been edited in the DB they will appear in the Dolibarr interface.
> 
> It has been two years since I last looked into this, but as I remember, I found a character filter in the PHP code that prevents typing, for example, å, ä, and ö. Back in 2023, its location was here: societe/card.php line 468

In Version 20.0
line 294 / 299 / 344 / 1002 / 1888
i believe it is the "filter_sanitize_email" and "filter_sanitize_url"

in line 344
	$object->url					= trim(GETPOST('url', 'custom', 0, FILTER_SANITIZE_URL));

i changed to

	$object->url					= GETPOST('url', 'alpha');

not nice, and manualy changing the database is no good as well, as soon as we change something else of the customer, the mail adress and url will be filtered again :-)

**Comment by hasse-basse** [comment]
This issue should be kept alive as the problem is still there.

**Comment by pixodeo** [comment]
Located the cause. It is in the GETPOST sanitization (core/lib/functions.lib.php, sanitizeVal), which is why the value is stripped on input but a value written directly in the DB displays fine.

Two masks are involved:

- `'email'` (used by `GETPOST('email', 'email')` on the thirdparty/contact forms):
  ```php
  case 'email':
      $out = filter_var($out, FILTER_SANITIZE_EMAIL);
  ```
  `FILTER_SANITIZE_EMAIL` drops every non-ASCII character, so an IDN address like `test@café.fr` becomes `test@caf.fr`.

- `'url'` (used by `GETPOST('url', 'url')`):
  ```php
  case 'url':
      $out = preg_replace('/[^:\/\[\]a-z0-9@\$\'\*\~\.\-_,;\?\!=%&+#]+/i', '', $out);
  ```
  That character class is ASCII-only, so the Unicode characters of an IDN host are removed as well.

So both fields silently mangle internationalized domain names at input time.

The fix is not a one-liner change to those masks, because sanitizeVal is a central security entry point (GETPOST) used everywhere, and simply widening the allowed characters to Unicode reintroduces homoglyph/spoofing concerns on those fields. The cleaner direction is to convert the host part to punycode with `idn_to_ascii()` (requires the intl extension) so the stored value stays ASCII and valid, rather than stripping. That is a design decision for the maintainers on a security-sensitive path, so I am flagging the exact spot and the options rather than pushing a change to the GETPOST filter. Happy to prepare a punycode-based patch if that direction is agreed.
